lib/beautiful/colorful.lua

```diff
@@ -0,0 +1,1395 @@
+-- @tparam number s Saturation float value 0-1
+-- @tparam number l Lightness float value 0-1
+-- @tparam number a Alpha float value 0-1
+-- @return A new color object, or nil if invalid
+-- @usage local c = colorful.color.from.hsla(30, 0.8, 0.6, 1)
+function color.from.hsla(h,s,l,a)
+    -- hue is adjustable as it's degrees
+    if not between(h, 0, 360) then
+        while h < 0 do
+            h = h + 360
+        end
+        while h > 360 do
+            h = h - 360
+        end
+    end
+    s = clamp(s, 0.0, 1.0)
+    l = clamp(l, 0.0, 1.0)
+    if s == nil or l == nil then return nil end
+    local r,g,b = hsl_to_rgb(h,s,l)
+    return create_color_object(r,g,b,h,s,l,a)
+end
+
+--- Create a new color instance from HSL values.
+-- Will call `color.from.hsla` with an Alpha value of 1
+-- @tparam number h Hue decimal value 0-360
+-- @tparam number s Saturation float value 0-1
+-- @tparam number l Lightness float value 0-1
+-- @return A new color object, or nil if invalid
+-- @usage local c = colorful.color.from.hsl(30, 0.8, 0.6)
+function color.from.hsl(h,s,l)
+    return color.from.hsla(h,s,l,1)
+end
+
+-- }}} HSL
+
+-- @section end
+
+-- }}} New Color functions
+
+-- {{{ Color transformation functions
+
+--- Color object transformation functions
+-- @section color_change
+
+--- Update the HSL values after editing any of the RGB values
+function color:hsl_update()
+    self.hue, self.saturation, self.lightness = rgb_to_hsl(self.red, self.green, self.blue)
+end
+
+--- Update the RGB values after editing any of the HSL values
+function color:rgb_update()
+    self.red, self.green, self.blue = hsl_to_rgb(self.hue, self.saturation, self.lightness)
+end
+
+-- {{{ RGB-adjusting functions
+
+-- Helper function for color:mix
+local function calcval(v,s,e)
+    local w = 1 - v
+    return s * v + e * w
+end
+
+--- Returns the mix of a color with another by a specified amount
+-- @tparam beautiful.colorful.color c A color object
+-- @tparam[opt=0.5] number v A number [0,1] of how much of the origin color
+-- should be mixed with c.
+-- @return A new color object representing the mix of the two colors, or nil if invalid
+-- @usage local mixed_color = c:mix(c2, 0.5)
+function color:mix(c, v)
+    if v == nil then v = 0.5 end
+    if v < 0 or v > 1 then return nil end
+    local red = calcval(v, self.red, c.red) * 255
+    local green = calcval(v, self.green, c.green) * 255
+    local blue = calcval(v, self.blue, c.blue) * 255
+    local alpha = calcval(v, self.alpha, c.alpha) * 255
+    return color.from.rgba(red, green, blue, alpha)
+end
+
+--- Returns the inverse of a color
+-- @return A new color object representing the inverted color, or nil if invalid
+-- @usage local inverted_color = c:invert()
+function color:invert()
+    local red = clamp(1 - self.red, 0, 1) * 255
+    local green = clamp(1 - self.green, 0, 1) * 255
+    local blue = clamp(1 - self.blue, 0, 1) * 255
+    return color.from.rgba(red, green, blue, self.alpha * 255)
```

Comment:
Why the `clamp`? If this actually changes anything, then the passed-in color was invalid. All the color-creation functions seem to be checking for that and clamping happens there.

Reply:
So this was in case somehow the red/green/blue values were updated to be outside `[0,1]`. Since the fields can be edited directly (right now), this was to prevent invalid colors from being created.